[익조] 스프링 카페 3단계 - DB에 저장하기

ikjo93 opened the pull request:
안녕하세요!! 익조입니다!! 😁

"스프링 카페 3단계 - DB에 저장하기" 구현 부분과 관련하여, 작성한 소스 코드 파일을 첨부드리오니 확인부탁드립니다. :)

배포 URL : https://ikjo-spring-cafe.herokuapp.com

※ 참고 : Heroku 배포 시에는 clear DB MySQL을 사용했습니다

아울러 미션 구현 세부 내용, 웹 앱의 이벤트 흐름도 등 미션 관련 사항은 별도 README 파일에 작성하였습니다.

### 고민사항
당초 1~2단계에서는 DB Layer에서의 객체와 View Layer에서의 객체간 역할 구분 없이 domain 객체가 모든 데이터 저장 및 비지니스 로직 등의 역할을 일괄 수행하고 있었습니다. 그리하여 객체의 역할과 책임 분리를 위해(+ 불필요한 데이터 교환 방지 위해) 3단계부터는 사용자로부터 입력을 전달받는 데이터 전송 객체인 Request Dto, 사용자에게 데이터를 응답해주는 데이터 전송 객체인 Response Dto를 추가하였습니다. 다만, 이 과정에서 Dto와 도메인 객체간 변환하는 작업이 처음이라 어려움이 많았습니다. 스택오버플로우 등의 자료를 참고하여 각 객체별로 변환 메서드를 선언하는 방식으로 변환 작업을 구현해보았는데, 사실 당당하지 못한 마음이 듭니다..😂 관련 자료를 찾는 중 ModelMapper 등 Dto와 도메인 객체간 변환 작업을 용이하게 해주는 라이브러리 등이 있었는데, 이번 미션에서는 JPA 뿐만 아니라 Lombok 등의 편의 기능도 제한하고 있기에 다소 취지에 맞지 않는 거 같아, 최대한 직접 구현해보려고 했었습니다. 다만 이 과정에서 미숙한 점이 많아 이 부분에 대한 리뷰도 해주시면 감사드리겠습니다 🙏

ikjo93 committed:
refactor : 2단계 코드리뷰 피드백 반영 및 코드 개선 작업

작업내용
1. (기존) 직접 자바 코드 작성을 통한 스프링 빈 등록 -> (변경) 컴포넌트 스캔 방식을 통한 스프링 빈 등록
  - WebAppConfig 클래스 파일 삭제
2. 도메인 클래스(User, Article)에서 도구를 이용하여 toStirng 메서드 오버라이드하도록 변경
3. User/ArticleController에서 logger.info 메서드 두번째 인자 toString 삭제(불필요한 문자열 연산)
4. ArticleController에서 PathVariable Integer -> int 변경(null 받는 경우 없음)
5. UserController에서 updateUser 핸들러 메서드에서 둘째 인자 변수명 updatedUser -> userDataUpdatedByUser로 정정
6. Article 객체 생성 시점에서 날짜 데이터 생성하도록 변경(LocalDate)
7. MemoryUserRepository에서 save 메서드 수정
 - synchronized 삭제 및 불필요한 예외처리문 삭제
 - findExistingUserId 메서드 삭제 등 기존 로직 개선(사용자 정보 생성 및 수정 기능 관련)
8. Optional 삭제(충분한 학습 이후 사용 예정)
9. 통합 테스트 코드 파일명 수정

ikjo93 committed:
refactor : 기존 도메인 객체 역할 분리 및 DTO 클래스(RequestDto, ResponseDto) 생성

작업내용
1. 기존 도메인 객체들의 경우 사용자로부터 입력을 받아오고 또한 사용자에게 데이터를 전달하는 등 계층간 데이터 교환하는 역할도 하고있었는데, 각 도메인별로 DTO 클래스(RequestDto, ResponseDto)를 생성하여 사용자와의 계층간 데이터 교환을 수행하는 역할은 별도 분리했습니다.
  - 이를 통해 도메인 객체들은 DB Layer(Repository)에서 데이터 처리 역할을 수행하고, DTO 객체들은 Presentation Layer에서 사용자와 데이터 교환을 하는 역할을 수행하도록 했습니다.
2. 이와 관련하여 기존 Controller, Service, Test 코드 등을 수정하였습니다.

ikjo93 committed:
feat : H2 데이터베이스 연동 및 사용자 정보 관련 기능 구현

작업내용
1. 사용자(User) 정보 저장, 수정, 조회 기능을 JDBC 템플릿을 이용해 H2 DB와 연동하였습니다.
2. 게시글 기능은 구현에 앞서 레파지토리 파일만 생성해놓은 상태입니다.

ikjo93 committed:
feat : H2 데이터베이스 연동 및 게시글 관련 기능 구현

작업내용 : 게시글(Article) 생성, 조회 기능을 JDBC 템플릿을 이용해 H2 DB와 연동하였습니다.

ikjo93 committed:
refactor : SQL문 별도 관리 및 통합 테스트 코드 파일 삭제

작업내용
1. 기존 레파지토리에 산재되어 있는 SQL문을 별도 클래스 파일에서 관리하도록 했습니다.

ikjo93 committed:
test : JdbcArticleRepository, JdbcUserRepository 단위 테스트 코드 생성

작업 내용
1. @JdbcTest를 활용하여 게시글 관련 DB 처리 담당 JdbcArticleRepository 계층과, 사용자 정보 관련 DB 처리 담당 JdbcUserRepository 계층의 단위 테스트 코드를 작성하였습니다.
2. 기존에 사용하던 BeanProperyRowMapper 대신 별도 RowMapper 메서드를 생성해주었습니다.(도메인 객체 setter 삭제)

ikjo93 committed:
docs : 미션 3단계 관련 README 파일 작성

이벤트 흐름도, 미션 구현 내용 등 작성

wheejuni commented on src/main/java/com/kakao/cafe/controller/ArticleController.java:
```diff
@@ -22,18 +22,17 @@ public ArticleController(ArticleService articleService) {
-        logger.info("POST /questions {}", article.toString());
-        articleService.upload(article);
+    public String createArticle(ArticleRequestDto articleRequestDto) {
+        logger.info("POST /questions {}", articleRequestDto);
```
`INFO` 레벨까지 부여해야 하는 로깅인지 잘 모르겠군요.
`INFO` 로 로깅되는 로그 메세지들이 어떤 것들인지는 스프링 앱 처음에 기동할때 콘솔에 뜨는 메세지들을 유심히 관찰해 보시면 되겠네요.

ikjo93 replied:
스프링 부트 앱을 처음 기동할 때 콘솔에 등장하는 로그들은 모두 INFO 레벨의 로그들이었습니다.

해당 내용을 살펴보면 웹 애플리케이션, 톰캣, 서블릿, 데이터 베이스 등의 초기화 동작에 관한 내용이 주를 이루었습니다.

또한 INFO 레벨에 대해 알아본 결과 INFO 레벨 수준의 로그들은 주로 "서비스의 시작, 구성, 중지 등 애플리케이션의 진행 과정(the progress of the application)을 강조하기 위해 사용된다고 합니다.

이러한 점들을 고려했을 때 말씀주신 것처럼 HTTP API, URL 등 사용자 요청 관련 정보는 INFO 레벨로 로깅하는 것은 적합하지 않다고 생각했습니다.

아울러, 현재 시점에서는 컨트롤러에서 모든 사용자의 요청에 대해 Logger를 통해 로그하고 있는 실정인데,

과연 이게 컨트롤러의 책임이 맞는지에 대한 의문이 생겼고 이와 관련하여 학습하는 중 핸들러 인터셉터에 대한 개념을 공부해야할 필요성을 느꼈습니다.

우선은 현재 단계에서 로그를 사용하기 보다는 추후에 로그 레벨에 대한 충분한 학습을 한 후 웹 앱에서 발생하는 일련의 이벤트들에 대해 다양한 레벨별로 체계적으로 로그를 관리할 수 있도록 해보겠습니다~🏃‍♂️

이번 미션을 진행 하면서 logger에 대해 관습적으로 사용하고 있었는데 좋은 피드백 감사드립니다.

또한 주신 피드백 덕분에 이번 기회에 로그 레벨에 대한 개념을 학습할 수 있었습니다.

wheejuni commented on src/main/java/com/kakao/cafe/domain/Article.java:
```diff
@@ -1,31 +1,33 @@
+    private LocalDateTime createdDate;
 
-    public Article(String writer, String title, String contents) {
+    private Article() {
+
+    }
```
불필요한 줄바꿈이 있는 듯합니다;

ikjo93 replied:
앗...! 넵!! 수정하겠습니다!! 😂

wheejuni commented on src/main/java/com/kakao/cafe/repository/JdbcArticleRepository.java:
```diff
@@ -0,0 +1,72 @@
+
+    private RowMapper<Article> articleRowMapper() {
+        return (rs, rowNum) -> {
+            Article article = new Article(rs.getInt("id"),
```
아이디가 `int` 인 것 괜찮을까요?

ikjo93 replied:
Integer는 int와 달리 Null로 초기화될 수 있는 장점이 있습니다. 그래도 해당 테이블(article)의 id 필드 값이 Null이라면 해당 ResultSe.getInt의 default value는 0으로 초기화되므로 DB로부터 Null 값을 가져오는 과정에서는 오류가 발생하진 않을 것 같지만

 혹여나 해당 테이블(article)의 id의 필드 값이 문자열로 저장되는 상황이 펼쳐진다면 이는 NumberFormatException가 발생할 우려가 있는데, 이를 예외 처리 해주면 오히려 코드가 더 복잡해질 것 같습니다.

물론 어떠한 오류도 발생하지 않는다고 가정하면 성능상 이점이 있는 int를 쓰는 것이 유리하겠지만, 항상 앞 일은 모르기 때문에 Integer를 사용하는 것이 적합할 것 같다는 생각이 들었습니다.

wheejuni commented on src/main/java/com/kakao/cafe/repository/JdbcArticleRepositorySqls.java:
```diff
@@ -0,0 +1,10 @@
+package com.kakao.cafe.repository;
+
+public class JdbcArticleRepositorySqls {
```
별로 권장하지 않습니다.
그냥 레포지토리에 하드코딩 되어있는 것이 더 유지보수 하기도 쉽고요, IDE가 제공하는 SQL 자동완성 같은 기능 사용하기도 쉽습니다.
사용처에 그냥 하드코딩해주세요~

ikjo93 replied:
아 그렇군요!! 감사합니다!! 수정하겠습니다!! 🏃‍♂️

wheejuni commented on src/main/java/com/kakao/cafe/repository/JdbcUserRepository.java:
```diff
@@ -0,0 +1,82 @@
+
+    @Override
+    public User save(User user) {
+        User userInformation;
+        try {
+            userInformation = findByUserId(user.getUserId())
+                .orElseThrow(() -> new NoSuchElementException("해당되는 ID가 없습니다."));
+        } catch (NoSuchElementException e) {
```
https://devlog-wjdrbs96.tistory.com/365 이걸 보고 한번 참고해보세요~

ikjo93 replied:
감사합니다!! 보내주신 링크 덕에 훨씬 간결한 코드를 작성할 수 있었습니다!! 🙏

wheejuni commented on src/main/java/com/kakao/cafe/repository/MemoryUserRepository.java:
```diff
@@ -1,45 +1,37 @@
-            users.set(index, user);
-        } catch (Exception e) {
-            return null;
+            userInformation = findByUserId(user.getUserId())
+                .orElseThrow(() -> new NoSuchElementException("해당되는 ID가 없습니다."));
+        } catch (NoSuchElementException e) {
+            users.add(user);
+            return user;
```
이 부분을 유연하게 풀 수 있는 다른 마무리 연산이 있을 것 같은데요?
`.orElse()` 나 `.orElseGet()` 같은 거요.

ikjo93 replied:
감사합니다!! 다음과 같이 orElse를 사용하여 개선할 수 있었습니다!! 😁

```Java
    @Override
    public User save(User user) {
        User userInformation = findByUserId(user.getUserId()).orElse(null);

        if (userInformation == null) {
            users.add(user);
            return user;
        }

        userInformation.update(user);

        return user;
    }
```

wheejuni reviewed (changes requested):
수고 많으셨습니다. 💯 
고민하신 부분은 별도 댓글로 답변 달게요~

wheejuni commented:
아 제가 특별히 추가 답변드릴 부분은 없긴 하네요.
그렇게 어색한 부분이 많이 보이진 않는데요? 제가 놓친 부분이 있다면 알려주세요.
코드리뷰를 요청하시기 전에 먼저 스스로 쭉 보시고 코멘트를 달아두시는 것도 좋은 방법입니다~

ikjo93 committed:
refactor : 3단계 피드백 반영 및 개선 작업

작업내용
1. 컨트롤러 로그 기능 삭제(추후 인터셉트를 통해 구현 예정)
2. 불필요한 줄바꿈 등 삭제
3. dto, rowmapper에서 int -> integer로 변경
4. 별도 sql 관리 파일 삭제 및 하드코딩
5. repository 저장 기능 로직 개선

ikjo93 commented:
@wheejuni
안녕하세요!! 그동안 Brian으로부터는 한 번도 리뷰를 받아 본 적이 없어서 이번에 기회가 닿아 리뷰 받을 수 있어서 좋았습니다!! 🥂

이번에 피드백 주신 부분은 방금 전 수정 및 push하였으니 확인 한 번 부탁드립니다~🏃‍♂️

마지막으로 이번에 Pull Request를 보냈던 내용 중에 미숙한 부분도 많았고 관습적으로 사용하고 있었던 부분들이 많았는데 잘 짚어주셔서 감사드렸습니다!! 덕분에 다소 거추장(?)스러웠던 코드들도 일부 간결하게 작성되어질 수 있었습니다! 🐱‍🏍

ikjo93 committed:
refactor : h2와 mysql 호환성 작업

honux77 commented:
수고하셨습니다. 요거 제가 머지할게요 ^^

ikjo93 commented:
> 수고하셨습니다. 요거 제가 머지할게요 ^^

@honux77
감사합니다 호눅스!! 5일동안 기다렸습니다 흑흑 😂🤣
